### PR TITLE
Constraint handling improvements

### PR DESCRIPTION
**Pull request by javdrher**
Fixes and improvements to the black-box support of the framework. Acquisition functions can now invalidate points and request which points are feasible.

This solves the issue reported in #10 which requires an acquisition to be aware of what points are feasible or not. As a consequence, the call to `setup()` from `set_data()` has to be delayed until all acquisition functions have updated data.

Note: some of the fixes assume highest_parent always ends in an acquisition function. For now this seems a safe assumption (?)

### CODE REVIEW

File: GPflowOpt/acquisition.py [hunk 1]
```diff
@@ -161,6 +172,10 @@ def __init__(self, model):
     def constraint_indices(self):
         return np.arange(self.data[1].shape[1])
 
+    def feasible_data_index(self):
+        pred = self.evaluate(self.data[0])
+        return pred.ravel() > 0.5
```
Comment (icouckuy) [inline GPflowOpt/acquisition.py]: K, so this is the main discussion point of this PR.

This is not really how things are done in literature (as far as I know). There are two things going on here:
- the feasibility of a data point is based on our belief and not on the actual (noisy) constraint value.
* For models with sn2 == 0 the prediction actually corresponds with the training data so no problem
* For models with sn2 > 0 this can be different. But when the data is noisy we perhaps don't want to use the noisy data either + if we know upfront there is noise we can fix sn2 to 0 (or jittery) during training.
-> all in all this seems OK.

- the check is normally done on the prediction itself (or the data) instead of the CDF + 0.5 seems a bit arbitrary.
* simply put: if the prediction is feasible (>0) it means the peak of the PDF is above >0 and because the Gaussian is non-skewed the CDF will always be >0.5 so that's OK too.

As discussed, maybe a doc string with one or two lines to inform the user (and us) would be nice: The feasibility of samples depends on our belief. For non-noisy models/data this corresponds to ... (standard in literature). ...

Comment (icouckuy) [inline GPflowOpt/acquisition.py]: I actually don't mind some feedback from the GPFlow people on what's the standard way to handle constraints or noise in machine learning.
